libraries/abstractions/pkcs11/psa/iot_pkcs11_psa_object_management.c

```diff
@@ -0,0 +1,600 @@
+* @param[in] ulDataSize	Size (in bytes) of data to be saved.
+* @param[in] pvContext	 Context of the key.
+*
+* @return The file handle of the object that was stored.
+*/
+CK_OBJECT_HANDLE PKCS11_PSA_SaveObject( CK_ATTRIBUTE_PTR pxClass,
+	CK_ATTRIBUTE_PTR pxLabel,
+	uint8_t * pucData,
+	uint32_t ulDataSize,
+	mbedtls_pk_context *pvContext )
+{
+	CK_OBJECT_HANDLE xHandle = eInvalidHandle;
+	psa_key_handle_t uxKeyHandle;
+	CK_ULONG uxKeyType;
+	psa_key_policy_t policy = psa_key_policy_init();
+	psa_algorithm_t uxAlgorithm;
+	CK_RV xBytesWritten = 0;
+	CK_RV xReturn;
+	uint32_t ulPresentMark = ( pkcs11OBJECT_PRESENT_MAGIC | ( ulDataSize ) );
+	uint8_t * pemBuffer;
+	size_t pemLength;
+	const psa_ps_create_flags_t uxFlags = PSA_PS_FLAG_NONE;
+	psa_ps_status_t uxStatus;
+	psa_ecc_curve_t curve_id;
+	const mbedtls_ecp_keypair *ec;
+	unsigned char *pcPrivateKeyDerStart = NULL;
+	size_t xPrivateKeySizeDer = 0;
+	unsigned char *pcPublicKeyUncompressedData = NULL;
+	size_t xPublicKeySizeUncompressed = 0;
+	uint8_t * pucKeyData = NULL;
+	uint32_t ulKeyDataSize = 0;
+	mbedtls_pk_type_t uxPrivateKeyTypePKCS11 = 0;
+
+	if( ulDataSize <= pkcs11OBJECT_MAX_SIZE )
+	{
+		/*
+		 * write client certificate.
+		 */
+		if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_CERTIFICATE ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_DEVICE_CERTIFICATE_FOR_TLS ) == 0 ) )
+		{
+			uxStatus = psa_ps_set( P11KeyConfig.uxDeviceCertificate,
+								   ulDataSize, pucData, uxFlags );
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulDeviceCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * write Jitp certificate.
+		 */
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_CERTIFICATE ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_JITP_CERTIFICATE ) == 0 ) )
+		{
+			uxStatus = psa_ps_set( P11KeyConfig.uxJitpCertificate,
+								   ulDataSize, pucData, uxFlags);
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulJitpCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * write Root certificate.
+		 */
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_CERTIFICATE ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_ROOT_CERTIFICATE ) == 0 ) )
+		{
+			uxStatus = psa_ps_set( P11KeyConfig.uxRootCertificate,
+								   ulDataSize, pucData, uxFlags);
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulRootCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * Store device private key by crypto service of PSA.
+		 */
+
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_PRIVATE_KEY ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS ) == 0 ) )
+		{
+			/*
+			 * Currently TF-M does not support persistent key APIs, so volatile
```

Comment:
Does this comment apply to the whole PKCS11 port herein, i.e., that persistent private keys aren't supported?

Reply:
Yes, the comment still applies. The persistent private keys have not been supported in TF-M yet. It is still under development now.